working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+var [a] = [1], (b, c) = (2, 3);
 ```
 
-#### Literal matcher
+Also, declarations with patterns must have an initializer. This is not a
+limitation since the point of using a pattern in a variable declaration is to
+match it against the initializer's value.
 
-A literal pattern determines if the value is equivalent to the given literal
-value.
+Add this new rule:
 
 ```
-literalMatcher ::=
-  | booleanLiteral
-  | nullLiteral
-  | numericLiteral
-  | stringLiteral
+patternDeclaration ::= ( 'final' | 'var' ) pattern '=' expression
 ```
 
-Note that list and map literals are not in here. Instead there are list and map
-*patterns*.
+It is a compile-time error if the outermost pattern in a `patternDeclaration`
+is not one of:
 
-**Breaking change**: Using matcher patterns in switch cases means that a list or
-map literal in a switch case is now interpreted as a list or map pattern which
-destructures its elements at runtime. Before, it was simply treated as identity
-comparison.
+*   [`listPattern`][listPattern]
+*   [`mapPattern`][mapPattern]
+*   [`recordPattern`][recordPattern]
+*   [`extractorPattern`][extractorPattern]
+*   [`nullCheckPattern`][nullCheckPattern]
 
-```dart
-const a = 1;
-const b = 2;
-var obj = [1, 2]; // Not const.
+This allows useful code like:
 
-switch (obj) {
-  case [a, b]: print("match"); break;
-  default: print("no match");
-}
+```dart
+var [a, b] = [1, 2];                 // List.
+var {1: a} = {1: 2};                 // Map.
+var (a, b, x: x) = (1, 2, x: 3);     // Record.
+var Point(x: x, y: y) = Point(1, 2); // Extractor.
+var string? = nullableString;        // Null-check.
 ```
 
-In Dart today, this prints "no match". With this proposal, it changes to
-"match". However, looking at the 22,943 switch cases in 1,000 pub packages
-(10,599,303 lines in 34,917 files), I found zero case expressions that were
-collection literals.
-
-#### Constant matcher
-
-Like literals, references to constants determine if the matched value is equal
-to the constant's value.
+But excludes other kinds of patterns to prohibit weird code like:
 
-```
-constantMatcher ::= qualified ( "." identifier )?
+```dart
+var String str = 'redundant';     // Variable.
+var str as String = 'weird';      // Cast.
+var definitely! = maybe;          // Null-assert.
+var (pointless) = 'parentheses';  // Grouping.
 ```
 
-The expression is syntactically restricted to be either:
+**TODO: Should we support destructuring in `const` declarations?**
 
-*   **A bare identifier.** In this case, the identifier must resolve to a
-    constant declaration in scope.
+This new rule is incorporated into the existing rules for declaring variables
+like so:
 
-*   **A prefixed or qualified identifier.** In other words, `a.b`. It must
-    resolve to either a top level constant imported from a library with a
-    prefix, a static constant in a class, or an enum case.
+```
+topLevelDeclaration ::=
+  | // Existing productions...
+  | 'late'? patternDeclaration ';' // New.
```

Comment:
I think we could make this work, but I agree it gets pretty weird. For now, I've gone ahead and removed support for late pattern variables. This also means removing support for patterns in top-level variables and static fields since those are also lazy-initialized. I think that's OK. We can always expand to support that in a future release.

Reply:
Removed support for late.